### PR TITLE
add support for Feature Extraction using PEFT

### PR DESCRIPTION
**Pull request by pacman100**
# What does this PR do?
1. For many tasks, users leverage `AutoModel` to get embeddings followed by postprocessing using that. One of the most popular use case of it is Semantic Similarity via Bi-Encoder models.
2. This PR enables that
3. Fixes https://github.com/huggingface/peft/issues/348
4. Adds an example on how to use it
5. Also shows how to use PEFT with custom models on top of Transformers

ToDo:
- [x] Add tests
- [x] Add docs
- [x] Add a complete example

### CODE REVIEW

File: src/peft/peft_model.py [hunk 1]
```diff
@@ -1545,3 +1545,62 @@ def _prefix_tuning_forward(
...
+                input_ids=input_ids,
+                attention_mask=attention_mask,
+                inputs_embeds=inputs_embeds,
+                output_attentions=output_attentions,
+                output_hidden_states=output_hidden_states,
+                return_dict=return_dict,
+                **kwargs,
+            )
+
+        batch_size = input_ids.shape[0]
```
Comment (BenjaminBossan) [inline src/peft/peft_model.py:1576]: It seems like `input_ids` can be `None`. Do we need to deal with that?

File: src/peft/peft_model.py [hunk 2]
```diff
@@ -1545,3 +1545,62 @@ def _prefix_tuning_forward(
...
+                "output_attentions": output_attentions,
+                "output_hidden_states": output_hidden_states,
+                "return_dict": return_dict,
+            }
+        )
+
+        if peft_config.peft_type == PeftType.PREFIX_TUNING:
+            past_key_values = self.get_prompt(batch_size)
+            return self.base_model(input_ids=input_ids, past_key_values=past_key_values, **kwargs)
+        else:
```
Comment (BenjaminBossan) [inline src/peft/peft_model.py:1600]: `else` is not strictly needed here, since the `if` clause returns, but it's also okay to leave it.

**Review by BenjaminBossan — COMMENTED** [review]
I'm still new to a lot of this, so I can't give a full review.

Here are some observations:

- I couldn't discern how the notebook `peft_prompt_tuning_seq2seq_with_generate.ipynb` relates to embeddings. Could this be elaborated? Also, it contains a big error message in the first cell.
- The notebook `peft_lora_embedding_semantic_similarity.ipynb` also contains an error message at the end. If it takes too long to run, could the number of epochs or the dataset size be reduced? Or at least the `KeyboardInterrupt` could be caught gracefully to avoid showing the error.
- In general, at least for me personally, it would be helpful if the notebooks could add a few sentences to explain what is going on and what the use case.
- There are no tests for the new class. Are we okay with that? Will it be added later?
- Similarly, there is also no documentation for the new class (doc entry, docstring), even though this seems to be a big addition.

**Comment by pacman100** [comment]
Hello @BenjaminBossan , the example is using PEFT Embedding Task to get embedding for queries and products and then uses them to get the cosine similarity between them to know whether the query and product are similar/relevant.

Also, the first cell didn't have any error, it is just using dataset from cache.

Yes, will be adding tests and docs.

**Comment by BenjaminBossan** [comment]
> the example is using PEFT Embedding Task to get embedding for queries and products and then uses them to get the cosine similarity between them to know whether the query and product are similar/relevant.
> 
> Also, the first cell didn't have any error, it is just using dataset from cache.

You mean in `peft_lora_embedding_semantic_similarity.ipynb`? Yes, but my comment referred to `peft_prompt_tuning_seq2seq_with_generate.ipynb`, which had a bunch of changes in this PR, and I was not sure how those relate to the PR.

> Yes, will be adding tests and docs.
> 
> Also working on a full example on this.

Fantastic. I'll do another review once they are added.

**Comment by pacman100** [comment]
> but my comment referred to peft_prompt_tuning_seq2seq_with_generate.ipynb, which had a bunch of changes in this PR, and I was not sure how those relate to the PR.

Those were due to `make style ` and `make quality`

**Comment by BenjaminBossan** [comment]
> Those were due to `make style ` and `make quality`

Ah I see, so just disregard my comments about that notebook.

File: docs/source/task_guides/semantic-similarity-lora.md [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+
+| Training Type | Training time per epoch (Hrs) | Batch Size that fits | ROC-AUC score (higher is better) |
+| ----------------- | ------------- | ---------- | -------- |
+| Pre-Trained e5-large-v2   |   -   | -          |  0.68    |
+| PEFT              |   1.73        | 64         |  0.787   |
+| Full Fine-Tuning  |   2.33        | 32         | 0.7969   |
+
+We can observe that PEFT-LoRA trains **1.35X** faster and can fit **2X** batch size when compared to full fine-tuning.
+The performance of PEFT-LoRA is comparable to full fine-tuning with relative drop of **-1.24%** in ROC-AUC. We believe this gap to close as the models get bigger as mentioned in [The Power of Scale for Parameter-Efficient Prompt Tuning
+](https://arxiv.org/pdf/2104.08691.pdf).
```
Comment (stevhliu) [inline docs/source/task_guides/semantic-similarity-lora.md:178]:
```suggestion
The PEFT-LoRA model trains **1.35X** faster and can fit **2X** batch size compared to the fully fine-tuned model, and the performance of PEFT-LoRA is comparable to the fully fine-tuned model with a relative drop of **-1.24%** in ROC-AUC. This gap can probably be closed with bigger models as mentioned in [The Power of Scale for Parameter-Efficient Prompt Tuning
](https://huggingface.co/papers/2104.08691).
```

File: docs/source/task_guides/semantic-similarity-lora.md [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+
+## Inference: Building search index and retrieving semantically similar products
```
Comment (stevhliu) [inline docs/source/task_guides/semantic-similarity-lora.md:180]:
```suggestion
## Inference
```

File: docs/source/task_guides/semantic-similarity-lora.md [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+
+Let's go! Now we have the model, we need to first create the search index of all the products in our catalog. 
+Please refer `peft_lora_embedding_semantic_similarity_inference.ipynb` for the entire inference code.
```
Comment (stevhliu) [inline docs/source/task_guides/semantic-similarity-lora.md:183]: Love the enthusiasm! Let's also add a link to the script when its merged :)

```suggestion
Let's go! Now we have the model, we need to create a search index of all the products in our catalog. 
Please refer to `peft_lora_embedding_semantic_similarity_inference.ipynb` for the complete inference code. 
```

File: docs/source/task_guides/semantic-similarity-lora.md [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+We can observe that PEFT-LoRA trains **1.35X** faster and can fit **2X** batch size when compared to full fine-tuning.
+The performance of PEFT-LoRA is comparable to full fine-tuning with relative drop of **-1.24%** in ROC-AUC. We believe this gap to close as the models get bigger as mentioned in [The Power of Scale for Parameter-Efficient Prompt Tuning
+](https://arxiv.org/pdf/2104.08691.pdf).
+
+## Inference: Building search index and retrieving semantically similar products
+
+Let's go! Now we have the model, we need to first create the search index of all the products in our catalog. 
+Please refer `peft_lora_embedding_semantic_similarity_inference.ipynb` for the entire inference code. 
+
+1. Get a list of ids to products. Let's call this `ids_to_products_dict`
```
Comment (stevhliu) [inline docs/source/task_guides/semantic-similarity-lora.md:185]:
```suggestion
1. Get a list of ids to products which we can call `ids_to_products_dict`:
```

File: docs/source/task_guides/semantic-similarity-lora.md [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+ 4: '(Set of 2) 15x6.00-6 Husqvarna/Poulan Tire Wheel Assy .75" Bearing',
+ 5: 'MaxAuto 2 Pcs 16x6.50-8 Lawn Mower Tire for Garden Tractors Ridings, 4PR, Tubeless',
+ 6: 'Dr.Roc Tire Spoon Lever Dirt Bike Lawn Mower Motorcycle Tire Changing Tools with Durable Bag 3 Tire Irons 2 Rim Protectors 1 Valve Stems Set TR412 TR413',
+ 7: 'MARASTAR 21446-2PK 15x6.00-6" Front Tire Assembly Replacement-Craftsman Mower, Pack of 2',
+ 8: '15x6.00-6" Front Tire Assembly Replacement for 100 and 300 Series John Deere Riding Mowers - 2 pack',
+ 9: 'Honda HRR Wheel Kit (2 Front 44710-VL0-L02ZB, 2 Back 42710-VE2-M02ZE)',
+ 10: 'Honda 42710-VE2-M02ZE (Replaces 42710-VE2-M01ZE) Lawn Mower Rear Wheel Set of 2' ...
+```
+
+2. Use out trained model to get the product embeddings. The model is available at [smangrul/peft_lora_e5_ecommerce_semantic_search_colab](https://huggingface.co/smangrul/peft_lora_e5_ecommerce_semantic_search_colab). Below is the relevant snippet:
```
Comment (stevhliu) [inline docs/source/task_guides/semantic-similarity-lora.md:201]:
```suggestion
2. Use the trained [smangrul/peft_lora_e5_ecommerce_semantic_search_colab](https://huggingface.co/smangrul/peft_lora_e5_ecommerce_semantic_search_colab) model to get the product embeddings:
```

File: docs/source/task_guides/semantic-similarity-lora.md [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+    with torch.no_grad():
+        with torch.amp.autocast(dtype=torch.bfloat16, device_type="cuda"):
+            product_embs = model(**{k:v.to(device) for k, v in batch.items()}).detach().float().cpu()
+    start_index = step*batch_size
+    end_index = start_index+batch_size if  (start_index+batch_size) < num_products else num_products
+    product_embeddings_array[start_index:end_index] = product_embs
+    del product_embs, batch
+```
+
+3. Create search index using HNSWlib
```
Comment (stevhliu) [inline docs/source/task_guides/semantic-similarity-lora.md:230]:
```suggestion
3. Create a search index using HNSWlib:
```

File: docs/source/task_guides/semantic-similarity-lora.md [hunk 9]
```diff
@@ -0,0 +1,306 @@
...
+    # Element insertion (can be called several times):
+    ids = np.arange(num_elements)
+    search_index.add_items(data, ids)
+
+    return search_index
+
+product_search_index = construct_search_index(d, num_products, product_embeddings_array)
+```
+
+4. utils for getting query embeddings and nearest neighbours
```
Comment (stevhliu) [inline docs/source/task_guides/semantic-similarity-lora.md:249]:
```suggestion
4. Get the query embeddings and nearest neighbors:
```

File: docs/source/task_guides/semantic-similarity-lora.md [hunk 10]
```diff
@@ -0,0 +1,306 @@
...
+    # Controlling the recall by setting ef:
+    search_index.set_ef(100) # ef should always be > k
+
+    # Query dataset, k - number of the closest elements (returns 2 numpy arrays)
+    labels, distances = search_index.knn_query(query_embeddings, k = k)
+    
+    return [(ids_to_products_dict[label], (1-distance)) for label, distance in zip(labels[0], distances[0]) if (1-distance)>=threshold]
+```
+
+5. Let's test it out with query `deep learning books`
```
Comment (stevhliu) [inline docs/source/task_guides/semantic-similarity-lora.md:270]:
```suggestion
5. Let's test it out with the query `deep learning books`:
```

File: docs/source/task_guides/semantic-similarity-lora.md [hunk 11]
```diff
@@ -0,0 +1,306 @@
...
+cosine_sim_score=0.9 product='Mastering Machine Learning on AWS: Advanced machine learning in Python using SageMaker, Apache Spark, and TensorFlow'
+cosine_sim_score=0.9 product='The Hundred-Page Machine Learning Book'
+cosine_sim_score=0.89 product='Hands-On Machine Learning with Scikit-Learn, Keras, and TensorFlow: Concepts, Tools, and Techniques to Build Intelligent Systems'
+cosine_sim_score=0.89 product='Machine Learning: A Journey from Beginner to Advanced Including Deep Learning, Scikit-learn and Tensorflow'
+cosine_sim_score=0.88 product='Mastering Machine Learning with scikit-learn'
+cosine_sim_score=0.88 product='Mastering Machine Learning with scikit-learn - Second Edition: Apply effective learning algorithms to real-world problems using scikit-learn'
+```
+
+we can see that books related to Deep Learning and Machine Learning are being retrieved even though the word `Machine Learning` doesn't appear in query. This means that model has learnt that these books are semantically relevant to the query based on purchase behaviour of customers of Amazon.
```
Comment (stevhliu) [inline docs/source/task_guides/semantic-similarity-lora.md:299]:
```suggestion
Books on deep learning and machine learning are retrieved even though `machine learning` wasn't included in the query. This means the model has learned that these books are semantically relevant to the query based on the purchase behavior of customers on Amazon. 
```

File: docs/source/task_guides/semantic-similarity-lora.md [hunk 12]
```diff
@@ -0,0 +1,306 @@
...
+cosine_sim_score=0.9 product='The Hundred-Page Machine Learning Book'
+cosine_sim_score=0.89 product='Hands-On Machine Learning with Scikit-Learn, Keras, and TensorFlow: Concepts, Tools, and Techniques to Build Intelligent Systems'
+cosine_sim_score=0.89 product='Machine Learning: A Journey from Beginner to Advanced Including Deep Learning, Scikit-learn and Tensorflow'
+cosine_sim_score=0.88 product='Mastering Machine Learning with scikit-learn'
+cosine_sim_score=0.88 product='Mastering Machine Learning with scikit-learn - Second Edition: Apply effective learning algorithms to real-world problems using scikit-learn'
+```
+
+we can see that books related to Deep Learning and Machine Learning are being retrieved even though the word `Machine Learning` doesn't appear in query. This means that model has learnt that these books are semantically relevant to the query based on purchase behaviour of customers of Amazon. 
+
+Next steps would ideally involve using ONNX/TensorRT to optimize the model and use triton server for hosting it. Check out 🤗 optimum for related optimizations for efficient serving.
```
Comment (stevhliu) [inline docs/source/task_guides/semantic-similarity-lora.md:301]:
```suggestion
The next steps would ideally involve using ONNX/TensorRT to optimize the model and using a Triton server to host it. Check out 🤗 [Optimum](https://huggingface.co/docs/optimum/index) for related optimizations for efficient serving!
```

File: src/peft/peft_model.py [hunk 13]
```diff
@@ -1545,3 +1545,62 @@ def _prefix_tuning_forward(
                 hidden_states=outputs.hidden_states,
                 attentions=outputs.attentions,
             )
+
+
+class PeftModelForFeatureExtraction(PeftModel):
```
Comment (stevhliu) [inline src/peft/peft_model.py:1550]: Would be awesome to include example usage in a code snippet like we currently have for the other `PeftModel`s!

**Review by stevhliu — COMMENTED** [review]
Thanks for adding a guide for such a cool use-case, this is great! 👏

Remember to add the docs to the `toctree` to properly build it!

**Comment by pacman100** [comment]
> Remember to add the docs to the toctree to properly build it!

Done

File: docs/source/task_guides/semantic-similarity-lora.md [hunk 14]
```diff
@@ -0,0 +1,300 @@
...
+We finetune [e5-large-v2](https://huggingface.co/intfloat/e5-large-v2) which tops the [MTEB benchmark](https://huggingface.co/spaces/mteb/leaderboard) using PEFT-LoRA. 
+
+[`AutoModelForSentenceEmbedding`] returns the query and product embeddings, and the `mean_pooling` function pools them across the sequence dimension and normalizes them:
+
+```py
+class AutoModelForSentenceEmbedding(nn.Module):
+    def __init__(self, model_name, tokenizer, normalize=True):
+        super(AutoModelForSentenceEmbedding, self).__init__()
+
+        self.model = AutoModel.from_pretrained(model_name)  # , load_in_8bit=True, device_map={"":0})
```
Comment (younesbelkada) [inline docs/source/task_guides/semantic-similarity-lora.md:123]:
```suggestion
        self.model = AutoModel.from_pretrained(model_name)  
```

File: src/peft/peft_model.py [hunk 15]
```diff
@@ -246,6 +246,8 @@ def _setup_prompt_encoder(self, adapter_name: str):
                 if transformer_backbone is None:
                     transformer_backbone = module
                     self.transformer_backbone_name = name
+        if transformer_backbone is None:
+            transformer_backbone = self.base_model
```
Comment (younesbelkada) [inline src/peft/peft_model.py:250]: Can you quickly explain why this change is needed? 🙏

**Review by younesbelkada — APPROVED** [review]
Very nice addition and learned a lot! I left 2 questions !

File: docs/source/task_guides/semantic-similarity-lora.md [hunk 16]
```diff
@@ -0,0 +1,300 @@
...
+
+💡 Read [LoRA: Low-Rank Adaptation of Large Language Models](https://arxiv.org/abs/2106.09685) to learn more about LoRA.
+
+</Tip>
+
+In this guide, we'll be using a LoRA [script](https://github.com/huggingface/peft/tree/main/examples/lora_dreambooth) to fine-tune a [`intfloat/e5-large-v2`](https://huggingface.co/intfloat/e5-large-v2) model on the [`smangrul/amazon_esci`](https://huggingface.co/datasets/smangrul/amazon_esci) dataset for semantic similarity tasks. Feel free to explore the script to learn how things work in greater detail!
+
+## Setup
+
+Start by installing 🤗 PEFT from [source](https://moon-ci-docs.huggingface.co/docs/peft/pr_647/en/install#source), and then navigate to the directory containing the training scripts for fine-tuning DreamBooth with LoRA:
```
Comment (BenjaminBossan) [inline docs/source/task_guides/semantic-similarity-lora.md:15]: Should this be linking to moon-ci? Also, is it necessary to install from source? We plan to have the release very soon.

File: docs/source/task_guides/semantic-similarity-lora.md [hunk 17]
```diff
@@ -0,0 +1,300 @@
...
+```
+
+2. Use the trained [smangrul/peft_lora_e5_ecommerce_semantic_search_colab](https://huggingface.co/smangrul/peft_lora_e5_ecommerce_semantic_search_colab) model to get the product embeddings:
+
+```py
+# base model
+model = AutoModelForSentenceEmbedding(model_name_or_path, tokenizer)
+
+# peft config and wrapping
+model = PeftModel.from_pretrained(model, peft_model_id)
```
Comment (BenjaminBossan) [inline docs/source/task_guides/semantic-similarity-lora.md:202]: Where is `peft_model_id` defined?

File: docs/source/task_guides/semantic-similarity-lora.md [hunk 18]
```diff
@@ -0,0 +1,300 @@
...
+</Tip>
+
+In this guide, we'll be using a LoRA [script](https://github.com/huggingface/peft/tree/main/examples/lora_dreambooth) to fine-tune a [`intfloat/e5-large-v2`](https://huggingface.co/intfloat/e5-large-v2) model on the [`smangrul/amazon_esci`](https://huggingface.co/datasets/smangrul/amazon_esci) dataset for semantic similarity tasks. Feel free to explore the script to learn how things work in greater detail!
+
+## Setup
+
+Start by installing 🤗 PEFT from [source](https://moon-ci-docs.huggingface.co/docs/peft/pr_647/en/install#source), and then navigate to the directory containing the training scripts for fine-tuning DreamBooth with LoRA:
+
+```bash
+cd peft/examples/embedding
```
Comment (BenjaminBossan) [inline docs/source/task_guides/semantic-similarity-lora.md:18]: Should this be 
```suggestion
cd peft/examples/feature_extraction
```
?

File: docs/source/task_guides/semantic-similarity-lora.md [hunk 19]
```diff
@@ -0,0 +1,300 @@
...
+- 🤗 Accelerate for the training loop
+- 🤗 Datasets for loading and preparing the `smangrul/amazon_esci` dataset for training and inference
+- 🤗 Evaluate for evaluating the model's performance
+- 🤗 PEFT for setting up the LoRA configuration and creating the PEFT model
+- 🤗 huggingface_hub for uploading the trained model to HF hub
+- hnswlib for creating the search index and doing fast approximate nearest neighbor search
+
+<Tip>
+
+It is assumed that PyTorch with CUDA support, Pandas, NumPy, and WandB are already installed.
```
Comment (BenjaminBossan) [inline docs/source/task_guides/semantic-similarity-lora.md:41]: Pandas (thus numpy) and WandB are part of the `requirements.txt`, so they don't need to be already installed.

File: docs/source/task_guides/semantic-similarity-lora.md [hunk 20]
```diff
@@ -0,0 +1,300 @@
...
+cosine_sim_score=0.89 product='Hands-On Machine Learning with Scikit-Learn, Keras, and TensorFlow: Concepts, Tools, and Techniques to Build Intelligent Systems'
+cosine_sim_score=0.89 product='Machine Learning: A Journey from Beginner to Advanced Including Deep Learning, Scikit-learn and Tensorflow'
+cosine_sim_score=0.88 product='Mastering Machine Learning with scikit-learn'
+cosine_sim_score=0.88 product='Mastering Machine Learning with scikit-learn - Second Edition: Apply effective learning algorithms to real-world problems using scikit-learn'
+```
+
+Books on deep learning and machine learning are retrieved even though `machine learning` wasn't included in the query. This means the model has learned that these books are semantically relevant to the query based on the purchase behavior of customers on Amazon. 
+
+The next steps would ideally involve using ONNX/TensorRT to optimize the model and using a Triton server to host it. Check out 🤗 [Optimum](https://huggingface.co/docs/optimum/index) for related optimizations for efficient serving!
+
```
Comment (BenjaminBossan) [inline docs/source/task_guides/semantic-similarity-lora.md:296]: Remove trailing lines.

**Review by BenjaminBossan — COMMENTED** [review]
I focused on the documentation part for this review, will review the code once the tests pass.

Great docs overall, thanks for putting so much work into this. I have a few minor comments. Please take a look.

**Review by younesbelkada — APPROVED** [review]
Still looking great, thanks !

**Review by BenjaminBossan — COMMENTED** [review]
Overall, this looks good to me, nice work.

I have one concern left, however. According to the coverage report. the following lines are not covered by tests:

https://github.com/huggingface/peft/blob/e2f7ff822cae7e1a23c5b7e7e5fac96781ec8f74/src/peft/peft_model.py#L1614-L1644

I double-checked locally and indeed I never hit those lines. When inspecting, only LoRA and IA³ are being tested. So should the tests be extended to include prompt learning or what could be done here?

**Comment by pacman100** [comment]
> I have one concern left, however. According to the coverage report. the following lines are not covered by tests:

latest commit should fix this

**Comment by AngledLuffa** [comment]
This is great!  Thanks for the support and for knocking off that old issue of mine.
